File: mobile/lib/models/gate_record.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class GateRecord {
  String id;
  String entryGate;
  DateTime entryScanTime;
  DateTime entryConfirmTime;
  String entryImageUrl;
  String exitGate;
  DateTime exitScanTime;
  DateTime exitConfirmTime;
  String exitImageUrl;
  String paymentStatus;
  DateTime paymentTime;
  String phoneNumber;
  String vehicleId;

  GateRecord({
    this.id,
    this.entryGate,
    this.entryScanTime,
    this.entryImageUrl,
    this.entryConfirmTime,
    this.exitGate,
    this.exitScanTime,
    this.exitImageUrl,
    this.exitConfirmTime,
    this.paymentStatus,
    this.paymentTime,
    this.phoneNumber,
    this.vehicleId
  });

  factory GateRecord.fromDocumentSnapshot(DocumentSnapshot doc) {
    return GateRecord(
      id: doc.documentID,
      entryGate: doc.data['entryGate'],
      entryScanTime: (doc.data['entryScanTime'] as Timestamp)?.toDate(),
      entryImageUrl: doc.data['entryImageUrl'],
      entryConfirmTime: (doc.data['entryConfirmTime'] as Timestamp)?.toDate(),
      exitGate: doc.data['exitGate'],
      exitScanTime: (doc.data['exitScanTime'] as Timestamp)?.toDate(),
      exitImageUrl: doc.data['entryImageUrl'],
      exitConfirmTime: (doc.data['exitConfirmTime'] as Timestamp)?.toDate(),
      paymentStatus: doc.data['paymentStatus'],
      paymentTime: (doc.data['paymentTime'] as Timestamp)?.toDate(),
      phoneNumber: doc.data['phoneNumber'],
      vehicleId: doc.data['vehicleId']
    );
  }
}

File: iot/src/gateRecord.ts
import * as firebase from "firebase/app"
import "firebase/firestore"

export enum GateMode {
    // eslint-disable-next-line no-unused-vars
    entry = 'entry',
    // eslint-disable-next-line no-unused-vars
    exit = 'exit'
}

export default class GateRecord {
    vehicleId: string = null
    phoneNumber: string = null
    entryScanTime: firebase.firestore.Timestamp = null
    entryConfirmTime: firebase.firestore.Timestamp = null
    entryImageUrl: string = null
    exitScanTime: firebase.firestore.Timestamp = null
    exitConfirmTime: firebase.firestore.Timestamp = null
    exitImageUrl: string = null
    entryGate: string = null
    exitGate: string = null
    paymentStatus: string = null
    paymentTime: firebase.firestore.Timestamp = null 
    
    constructor(deviceId: string, gateMode: GateMode, vehicleId: string, imageUrl: string) {
        this.vehicleId = vehicleId
        switch (gateMode) {
            case GateMode.entry:
                this.entryGate = deviceId
                this.entryScanTime = firebase.firestore.Timestamp.fromDate(new Date())
                this.entryImageUrl = imageUrl
                break
            case GateMode.exit:
                this.exitGate = deviceId
                this.exitScanTime = firebase.firestore.Timestamp.fromDate(new Date())
                this.exitImageUrl = imageUrl
                break
        }
    }
}
